src/order/bounds.lean

```diff
@@ -1,378 +1,609 @@
 def is_greatest (s : set α) (a : α) : Prop := a ∈ s ∧ a ∈ upper_bounds s
+
 /-- `a` is a least upper bound of a set `s`; for a partial order, it is unique if exists. -/
 def is_lub (s : set α) : α → Prop := is_least (upper_bounds s)
 /-- `a` is a greatest lower bound of a set `s`; for a partial order, it is unique if exists. -/
 def is_glb (s : set α) : α → Prop := is_greatest (lower_bounds s)
 
-lemma upper_bounds_mono (h₁ : a₁ ≤ a₂) (h₂ : a₁ ∈ upper_bounds s) : a₂ ∈ upper_bounds s :=
-λ a h, le_trans (h₂ h) h₁
+/-!
+### Monotonicity
+-/
 
-lemma lower_bounds_mono (h₁ : a₂ ≤ a₁) (h₂ : a₁ ∈ lower_bounds s) : a₂ ∈ lower_bounds s :=
-λ a h, le_trans h₁ (h₂ h)
+lemma upper_bounds_mono_set ⦃s t : set α⦄ (hst : s ⊆ t) :
+  upper_bounds t ⊆ upper_bounds s :=
+λ b hb x h, hb $ hst h
 
-lemma mem_upper_bounds_image (Hf : monotone f) (Ha : a ∈ upper_bounds s) :
-  f a ∈ upper_bounds (f '' s) :=
-ball_image_of_ball (assume x H, Hf (Ha ‹x ∈ s›))
+lemma lower_bounds_mono_set ⦃s t : set α⦄ (hst : s ⊆ t) :
+  lower_bounds t ⊆ lower_bounds s :=
+λ b hb x h, hb $ hst h
 
-lemma mem_lower_bounds_image (Hf : monotone f) (Ha : a ∈ lower_bounds s) :
-  f a ∈ lower_bounds (f '' s) :=
-ball_image_of_ball (assume x H, Hf (Ha ‹x ∈ s›))
+lemma upper_bounds_mono_mem ⦃a b⦄ (hab : a ≤ b) : a ∈ upper_bounds s → b ∈ upper_bounds s :=
+λ ha x h, le_trans (ha h) hab
 
-lemma is_lub_singleton {a : α} : is_lub {a} a :=
-by simp [is_lub, is_least, upper_bounds, lower_bounds] {contextual := tt}
+lemma lower_bounds_mono_mem ⦃a b⦄ (hab : a ≤ b) : b ∈ lower_bounds s → a ∈ lower_bounds s :=
+λ hb x h, le_trans hab (hb h)
 
-lemma is_glb_singleton {a : α} : is_glb {a} a :=
-by simp [is_glb, is_greatest, upper_bounds, lower_bounds] {contextual := tt}
+lemma upper_bounds_mono ⦃s t : set α⦄ (hst : s ⊆ t) ⦃a b⦄ (hab : a ≤ b) :
+  a ∈ upper_bounds t → b ∈ upper_bounds s :=
+λ ha, upper_bounds_mono_set hst $ upper_bounds_mono_mem hab ha
 
-lemma is_glb_Ici : is_glb (Ici a) a :=
-⟨λx hx, hx, λy hy, hy left_mem_Ici⟩
+lemma lower_bounds_mono ⦃s t : set α⦄ (hst : s ⊆ t) ⦃a b⦄ (hab : a ≤ b) :
+  b ∈ lower_bounds t → a ∈ lower_bounds s :=
+λ hb, lower_bounds_mono_set hst $ lower_bounds_mono_mem hab hb
 
-lemma is_glb_Icc (h : a₁ ≤ a₂) : is_glb (Icc a₁ a₂) a₁ :=
-⟨λx hx, hx.1, λy hy, hy (left_mem_Icc.mpr h)⟩
+/-- If `s ⊆ t` and `t` is bounded above, then so is `s`. -/
+lemma bdd_above.mono ⦃s t : set α⦄ (h : s ⊆ t) : bdd_above t → bdd_above s :=
+nonempty.mono $ upper_bounds_mono_set h
 
-lemma is_glb_Ico (h : a₁ < a₂) : is_glb (Ico a₁ a₂) a₁ :=
-⟨λx hx, hx.1, λy hy, hy (left_mem_Ico.mpr h)⟩
+/-- If `s ⊆ t` and `t` is bounded below, then so is `s`. -/
+lemma bdd_below.mono ⦃s t : set α⦄ (h : s ⊆ t) : bdd_below t → bdd_below s :=
+nonempty.mono $ lower_bounds_mono_set h
 
-lemma is_lub_Iic : is_lub (Iic a) a :=
-⟨λx hx, hx, λy hy, hy right_mem_Iic⟩
+lemma is_lub.squeeze {s t p : set α} (hs : is_lub s a) (hp : is_lub p a)
```

Comment:
I don't really like the name because I think it is really hard to guess. Could you at least add a docstring? Same thing for `glb`.

Reply:
Any better suggestions?

Reply:
ìs_lub.of_subset_of_superset`?